spec/unit/environments_spec.rb

```diff
@@ -607,6 +607,51 @@
         cached.get(:cached)
       end
 
+      it "normalizes environment name to symbol" do
+        env = Puppet::Node::Environment.create(:cached, [])
+        mocked_loader = double('loader')
+
+        expect(mocked_loader).not_to receive(:get).with('cached')
+        expect(mocked_loader).to receive(:get).with(:cached).and_return(env).once
+        expect(mocked_loader).to receive(:get_conf).with(:cached).and_return(Puppet::Settings::EnvironmentConf.static_for(env, 20)).once
+
+        cached = Puppet::Environments::Cached.new(mocked_loader)
+        cached.get('cached')
+        cached.get(:cached)
+      end
+
+      it "caches environment name as symbol and only once" do
+        mocked_loader = double('loader')
+
+        env = Puppet::Node::Environment.create(:cached, [])
+        allow(mocked_loader).to receive(:get).with(:cached).and_return(env)
+        allow(mocked_loader).to receive(:get_conf).with(:cached).and_return(Puppet::Settings::EnvironmentConf.static_for(env, 20))
+
+        cached = Puppet::Environments::Cached.new(mocked_loader)
+        cached.get(:cached)
+        cached.get('cached')
+
+        expect(cached.instance_variable_get(:@cache).keys).to eq([:cached])
```

Comment:
Instead of peeking at instance variables, maybe use the `environment` matcher defined at the bottom of this file?

```ruby
expect(loader.list).to contain_exactly(environment(:cached))
```

Reply:
I don't think that the `list` method of `Puppet::Environments::Cached` can be used to test this as it doesn't seem broken and is not affected by the fix. Running below script (without fix):
```ruby
require 'puppet'
Puppet.initialize_settings
Puppet[:log_level] = 'debug'
Puppet::Util::Log.newdestination(:console)

Puppet::ApplicationSupport.push_application_context(Puppet::Util::RunMode[:user])

envs = Puppet.lookup(:environments)
pp envs.instance_variable_get(:@cache).keys
puts
envs.get!('production')
pp envs.instance_variable_get(:@cache).keys
puts
envs.get!(:production)
pp envs.instance_variable_get(:@cache).keys
puts
pp envs
puts
pp envs.list
```
gives below output:
```sh-session
➜ bundle exec ruby script.rb
Debug: Caching environment 'production' (ttl = 1 sec)
["production"]

Debug: Found in cache 'production' (ttl = 1 sec)
["production"]

Debug: Caching environment 'production' (ttl = 1 sec)
["production", :production]

#<Puppet::Environments::Cached:0x00007ff87c671570
 @cache=
  {"production"=>
    #<Puppet::Environments::Cached::TTLEntry:0x00007ff87c6aaa28
     @ttl=2021-06-24 16:06:19 +0300,
     @ttl_seconds=1,
     @value=
      <Puppet::Node::Environment:70352607937960 @name="production" @manifest="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/manifests" @modulepath="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/modules:/Users/luchian.nemes/.puppetlabs/etc/code/modules:/opt/puppetlabs/puppet/modules" >>,
   :production=>
    #<Puppet::Environments::Cached::TTLEntry:0x00007ff87c7a9370
     @ttl=2021-06-24 16:06:19 +0300,
     @ttl_seconds=1,
     @value=
      <Puppet::Node::Environment:70352608516880 @name="production" @manifest="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/manifests" @modulepath="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/modules:/Users/luchian.nemes/.puppetlabs/etc/code/modules:/opt/puppetlabs/puppet/modules" >>},
 @cache_expiration_service=
  #<Puppet::Environments::Cached::DefaultCacheExpirationService:0x00007ff87c6714a8>,
 @loader=
  #<Puppet::Environments::Combined:0x00007ff87c671638
   @loaders=
    [#<Puppet::Environments::Directories:0x00007ff87c672f60
      @environment_dir=
       "/Users/luchian.nemes/.puppetlabs/etc/code/environments",
      @global_module_path=
       ["/Users/luchian.nemes/.puppetlabs/etc/code/modules",
        "/opt/puppetlabs/puppet/modules"]>,
     #<Puppet::Environments::StaticPrivate:0x00007ff87c671a20
      @environments=
       [<Puppet::Node::Environment:70352607875000 @name="production" @manifest="no_manifest" @modulepath="/Users/luchian.nemes/.puppetlabs/etc/code/modules:/opt/puppetlabs/puppet/modules" >]>]>>

[<Puppet::Node::Environment:70352608516880 @name="production" @manifest="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/manifests" @modulepath="/Users/luchian.nemes/.puppetlabs/etc/code/environments/production/modules:/Users/luchian.nemes/.puppetlabs/etc/code/modules:/opt/puppetlabs/puppet/modules" >]
```
The environment appears only once when running `envs.list`, even though it’s clearly cached twice.

Reply:
Oh, this is a bug in the environment cache! The [`list`](https://github.com/puppetlabs/puppet/blob/aa5e3fdab6fb63f34fcfc36982e69b5d4e7a0575/lib/puppet/environments.rb#L359) method always returns environments based on what's in the filesystem due to https://github.com/puppetlabs/puppet/blob/aa5e3fdab6fb63f34fcfc36982e69b5d4e7a0575/lib/puppet/environments.rb#L254 So that's why we only get one entry, even though there are two entries in the cache.

One option might be to clear entries that are in the cache, but no longer on disk:

```ruby
      current_envs = @loader.list
      @cache.keys - current_envs.each do |deleted|
        Puppet.debug { "Environment no longer exists '#{deleted}'"}
        clear(deleted)
      end

      current_envs.map do |env|
        name = env.name
        old_entry = @cache[name]
        ...
```

Reply:
I'm ok calling `expect(cached.instance_variable_get(:@cache).keys).to eq([:cached])` in this PR. I filed [PUP-11129](https://tickets.puppetlabs.com/browse/PUP-11129) about the bug above.

Reply:
Cool, I've added a solution as comment to the ticket. (I've had some trouble testing it with rspec though).